libraries/persistent_store/src/model.rs

```diff
@@ -0,0 +1,164 @@
+#[derive(Clone, Debug)]
+pub struct StoreModel {
+    /// Represents the content of the store.
+    map: HashMap<usize, Box<[u8]>>,
+
+    /// The modeled storage configuration.
+    format: Format,
+}
+
+/// Mutable operations on a store.
+#[derive(Clone, Debug)]
+pub enum StoreOperation {
+    /// Applies a transaction.
+    Transaction {
+        /// The list of updates to be applied.
+        updates: Vec<StoreUpdate>,
+    },
+
+    /// Deletes all keys above a threshold.
+    Clear {
+        /// The minimum key to be deleted.
+        min_key: usize,
+    },
+
+    /// Compacts the store until a given capacity is immediately available.
+    Prepare {
+        /// How much capacity should be immediately available after compaction.
+        length: usize,
+    },
+}
+
+impl StoreModel {
+    /// Creates an empty model for a given storage configuration.
+    pub fn new(format: Format) -> StoreModel {
+        let map = HashMap::new();
+        StoreModel { map, format }
+    }
+
+    /// Returns the modeled content.
+    pub fn map(&self) -> &HashMap<usize, Box<[u8]>> {
+        &self.map
+    }
+
+    /// Returns the storage configuration.
+    pub fn format(&self) -> &Format {
+        &self.format
+    }
+
+    /// Simulates a store operation.
+    pub fn apply(&mut self, operation: StoreOperation) -> StoreResult<()> {
+        match operation {
+            StoreOperation::Transaction { updates } => self.transaction(updates),
+            StoreOperation::Clear { min_key } => self.clear(min_key),
+            StoreOperation::Prepare { length } => self.prepare(length),
+        }
+    }
+
+    /// Returns the capacity according to the model.
+    pub fn capacity(&self) -> StoreRatio {
+        let total = self.format.total_capacity();
+        let used: usize = self.map.values().map(|x| self.entry_size(x)).sum();
+        StoreRatio { used, total }
+    }
+
+    /// Applies a transaction.
+    fn transaction(&mut self, updates: Vec<StoreUpdate>) -> StoreResult<()> {
+        // Fail if too many updates.
+        if updates.len() > self.format.max_updates() {
+            return Err(StoreError::InvalidArgument);
+        }
+        // Fail if an update is invalid.
+        if !updates.iter().all(|x| self.update_valid(x)) {
+            return Err(StoreError::InvalidArgument);
+        }
+        // Fail if updates are not disjoint, i.e. there are duplicate keys.
+        let keys: HashSet<_> = updates.iter().map(|x| x.key()).collect();
+        if keys.len() != updates.len() {
+            return Err(StoreError::InvalidArgument);
+        }
+        // Fail if there is not enough capacity.
+        let capacity = match updates.len() {
+            0 => 0,
+            1 => match &updates[0] {
+                StoreUpdate::Insert { value, .. } => self.entry_size(value),
+                StoreUpdate::Remove { .. } => 0,
+            },
+            _ => 1 + updates.iter().map(|x| self.update_size(x)).sum::<usize>(),
```

Comment:
This refers to a special property of the store, maybe document / link / mention something?

Reply:
Done.